### PR TITLE
Arbitrary approx order

### PR DESCRIPTION
**Pull request by xtalax**
Generalizes approximation order, allowing arbitrary order finite differences.

- [x] Move boundary conditions over to a rules based implementation.
- [x] Cartesian Finite Difference.
- [x] Spherical Finite Difference. REDUNDANT
- [x] Nonlinear Cartesian Laplacian.
- [x] Nonlinear Spherical Laplacian.
- [x] Component Tests.
- [x] Integration Tests.
- [x] Tests passing

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -4,3 +4,9 @@
 Manifest.toml
 .vscode
 .vscode/*
+plots
+plots/*
```
Comment (ChrisRackauckas) [inline .gitignore:8]: ?

File: src/MOLFiniteDifference.jl [hunk 2]
```diff
@@ -0,0 +1,15 @@
+struct MOLFiniteDifference{G} <: DiffEqBase.AbstractDiscretization
+    dxs
+    time
+    approx_order::Int
+    grid_align::G
+end
+
+# Constructors. If no order is specified, both upwind and centered differences will be 2nd order
```
Comment (ChrisRackauckas) [inline src/MOLFiniteDifference.jl:8]: upwind should default to first order.

File: src/MOLFiniteDifference.jl [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+function MOLFiniteDifference(dxs, time=nothing; approx_order = 2, grid_align=CenterAlignedGrid())
```
Comment (ChrisRackauckas) [inline src/MOLFiniteDifference.jl:9]: why centered default?

File: src/bcs/generate_bc_eqs.jl [hunk 4]
```diff
@@ -1,119 +1,189 @@
...
+        # * Assume in the form `u(...) ~ ...` for now
         bcdepvar = first(get_depvars(bc.lhs, depvar_ops))
-        if any(u->isequal(operation(u),operation(bcdepvar)), s.vars)
+        
+        if any(u -> isequal(operation(u), operation(bcdepvar)), s.ū)
             if t !== nothing && operation(bc.lhs) isa Sym && !any(x -> isequal(x, t.val), arguments(bc.lhs))
                 # initial condition
-                # Assume in the form `u(...) ~ ...` for now
-                initindex = findfirst(isequal(bc.lhs), boundarymap.initial) 
+                # * Assume that the initial condition is not in terms of the initial derivative
```
Comment (ChrisRackauckas) [inline src/bcs/generate_bc_eqs.jl:83]: this comment is not related to the one before? The code looks like it has the same limitation?

File: src/bcs/boundary_types.jl [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+abstract type AbstractTruncatingBoundary <: AbstractBoundary end
+
+abstract type AbstractExtendingBoundary <: AbstractBoundary end
+
+struct LowerBoundary <: AbstractTruncatingBoundary
+end
+
+struct UpperBoundary<: AbstractTruncatingBoundary
+end
```
Comment (ChrisRackauckas) [inline src/bcs/boundary_types.jl:13]: What does this mean for 3D?

File: src/discretization/generate_finite_difference_rules_old.jl [hunk 6]
```diff
@@ -0,0 +1,128 @@
+function calculate_weights_cartesian(order::Int, x0::T, xs::AbstractVector, idxs::AbstractVector) where T<:Real
```
Comment (ChrisRackauckas) [inline src/discretization/generate_finite_difference_rules_old.jl]: why is this file around?

File: .gitignore (hunk 1 above)
Comment (xtalax) [reply .gitignore:8]: That should't be there, remobng

File: test/MOL_original.jl [hunk 7]
```diff
@@ -0,0 +1,416 @@
+using ModelingToolkit: operation, istree, arguments, variable
```
Comment (ChrisRackauckas) [inline test/MOL_original.jl:1]: what's this for?

File: src/MOLFiniteDifference.jl (hunk 3 above)
Comment (xtalax) [reply src/MOLFiniteDifference.jl:9]: It was previously

File: test/pde_systems/MOL_1D_NonLinear_Diffusion.jl [hunk 8]
```diff
@@ -59,11 +59,67 @@ using ModelingToolkit: Differential
     #using Plots
     #plot(r_space, asf, seriestype = :scatter,label="analytic solution")
     #plot!(r_space, sol′, label="numeric solution")
-    #savefig("MOL_NonLinear_Diffusion_1D_Test00.png")
+    #savefig("plots/MOL_NonLinear_Diffusion_1D_Test00.png")
 
 end
 
-@testset "Test 01: Dt(u(t,x)) ~ Dx(u(t,x)^2 * Dx(u(t,x)))" begin
+# @testset "Test 01: Dt(u(t,x)) ~ Dx(u(t,x) * Dx(u(t,x)))" begin
```
Comment (ChrisRackauckas) [inline test/pde_systems/MOL_1D_NonLinear_Diffusion.jl:66]: please do not comment out tests. Always just mark them as broken instead.

File: src/MOLFiniteDifference.jl (hunk 3 above)
Comment (ChrisRackauckas) [reply src/MOLFiniteDifference.jl:9]: it only defaulted to centered when Neumann IIRC?

File: README.md [hunk 9]
```diff
@@ -9,24 +9,24 @@ Feature requests and issues welcome.
...
+                                      approx_order = Order of derivative approximation, starting from 2 
                                       grid_align = your grid type choice>)
 prob = discretize(pdesys, discretization)
 ```
 Where `dxs` is a vector of pairs of parameters to the grid step in this dimension, i.e. `[x=>0.2, y=>0.1]`
 
 Note that the second argument to `MOLFiniteDifference` is optional, all parameters can be discretized if all boundary conditions are specified
 
-Currently supported grid types: `center_align` and `edge_align`
+Currently supported grid types: `center_align` and `edge_align`. Edge align will give better accuracy with Neumann Boundary conditions.
```
Comment (ChrisRackauckas) [inline README.md:20]: should be the other way around?

**Comment by ChrisRackauckas** [comment]
This has gotten far too long, so I think it's good to just merge and clean in smaller PRs from here.
